agent/lib/kontena/rpc_client.rb

```diff
@@ -52,21 +53,28 @@ def notification(method, params)
     # @raise abort
     # @return [Object]
     def request(method, params, timeout: 30)
-      id = request_id
-      @requests[id] = nil
-
       if !wait_until("websocket client is connected", timeout: timeout, threshold: 10.0, interval: 0.1) { connected? }
         raise TimeoutError.new(500, 'WebsocketClient is not connected')
       end
 
+      id = request_id
+      cond = Celluloid::Condition.new
+      @requests[id] = cond
+
+      start = Time.now
       websocket_client.send_request(id, method, params)
 
-      if !wait_until("request #{method} has response wth id=#{id}", timeout: timeout, interval: 0.01) { @requests[id] }
+      begin
+        cond.wait(timeout)
```

Comment:
Background thought here: implementing some kind of `Observable#wait` API to replace the `Celluloid::Condition` would be a much better solution here :)

* Avoid the dual use of the `@requests[id]` for storing both a `Celluloid::Condition` and `[result, error]` - it would just be an `Observable` with an `observable_value`
* Avoid the theoretical race condition by having the `wait` return immediately once the observable value is updated